### PR TITLE
Support Map type in ORC extension

### PR DESCRIPTION
**Pull request by dansuzuki**
This is in reference to #5362

### CODE REVIEW

**Comment by jon-wei** [comment]
@dansuzuki Can you rebase your PR branch to the current druid master? Looks like there are some issues/conflicts related to that.

**Comment by dansuzuki** [comment]
Alright, will do.

File: extensions-contrib/orc-extensions/src/main/java/io/druid/data/input/orc/OrcHadoopInputRowParser.java [hunk 1]
```diff
@@ -132,6 +138,23 @@ private List getListObject(ListObjectInspector listObjectInspector, Object listO
...
+  private void getMapObject(String parentName, MapObjectInspector mapObjectInspector, Object mapObject, Map<String, Object> parsedMap)
+  {
+    if (mapObjectInspector.getMapSize(mapObject) < 0) {
+      return;
+    }
+    String prefix = parentName + "_";
+    Map objectMap = mapObjectInspector.getMap(mapObject);
+    PrimitiveObjectInspector key = (PrimitiveObjectInspector) mapObjectInspector.getMapKeyObjectInspector();
+    PrimitiveObjectInspector value = (PrimitiveObjectInspector) mapObjectInspector.getMapValueObjectInspector();
+    for (Iterator itr = objectMap.entrySet().iterator(); itr.hasNext(); ) {
```
Comment (jon-wei) [inline extensions-contrib/orc-extensions/src/main/java/io/druid/data/input/orc/OrcHadoopInputRowParser.java]: There's a TeamCity inspection error here:

```
 extensions-contrib/orc-extensions/src/main/java/io/druid/data/input/orc
 OrcHadoopInputRowParser.java (1)
150: getMapObject() for loop replaceable with 'foreach'
```

Comment (dansuzuki) [reply extensions-contrib/orc-extensions/src/main/java/io/druid/data/input/orc/OrcHadoopInputRowParser.java]: @jon-wei it's already fixed.

File: extensions-contrib/orc-extensions/src/main/java/io/druid/data/input/orc/OrcHadoopInputRowParser.java [hunk 2]
```diff
@@ -132,6 +137,22 @@ private List getListObject(ListObjectInspector listObjectInspector, Object listO
     return list;
   }
 
+  private void getMapObject(String parentName, MapObjectInspector mapObjectInspector, Object mapObject, Map<String, Object> parsedMap)
+  {
+    if (mapObjectInspector.getMapSize(mapObject) < 0) {
+      return;
+    }
+    String prefix = parentName + "_";
```
Comment (jon-wei) [inline extensions-contrib/orc-extensions/src/main/java/io/druid/data/input/orc/OrcHadoopInputRowParser.java]: What do you think about allowing the user to provide a format string that defines how the map field name and the subkey names are combined? Could be useful for avoiding potential field name collisions.

Comment (dansuzuki) [reply extensions-contrib/orc-extensions/src/main/java/io/druid/data/input/orc/OrcHadoopInputRowParser.java]: Definitely would be useful. I could think of single handling for all map-types for now though. E.g.
```
"parser" : {
  "mapTypeFieldNameResolver": "{PARENT}_{KEY}"
}
```
Would you agree?

Comment (jon-wei) [reply extensions-contrib/orc-extensions/src/main/java/io/druid/data/input/orc/OrcHadoopInputRowParser.java]: Sounds good, I think a single format for all map types is fine

**Comment by jihoonson** [comment]
@dansuzuki does this patch support nested Map types as well? If not, please leave a comment somewhere.

**Comment by dansuzuki** [comment]
@jihoonson no, not yet. For now, I can't think of not-so-crude way to resolve column names for nested map types.

File: docs/content/development/extensions-contrib/orc.md [hunk 3]
```diff
@@ -17,10 +17,11 @@ This is for batch ingestion using the HadoopDruidIndexer. The inputFormat of inp
...
 |typeString| String      | String representation of Orc struct type info. If not specified, auto constructed from parseSpec but all metric columns are dropped | no|
+|mapFieldNameFormat| String | String format for resolving the flatten map fields. Default is `<PARENT>_<CHILD>`. | no |
 
 For example of `typeString`, string column col1 and array of string column col2 is represented by `"struct<col1:string,col2:array<string>>"`.
 
-Currently, it only supports java primitive types and array of java primitive types, which means only 'list' of compound types in [ORC types](https://orc.apache.org/docs/types.html) is supported (list of list is not supported).  
+Currently, it only supports java primitive types, array of java primitive types and map of java primitive types. Thus, compound types 'list' and 'map' in [ORC types](https://orc.apache.org/docs/types.html) are supported. Note that, list of list is not supported, nor map of compond types. For map types, values will be exploded to several columns where column names will be resolved via `mapFieldNameFormat`.
```
Comment (jihoonson) [inline docs/content/development/extensions-contrib/orc.md]: `compond` -> `compound`

**Comment by jihoonson** [comment]
@dansuzuki ok. I think we should improve Druid's schema system to fundamentally support nested types in the future.

**Comment by jon-wei** [comment]
@dansuzuki Can you please sign the CLA? http://druid.io/community/cla.html

**Comment by dansuzuki** [comment]
@jon-wei just signed it. Thanks!

**Pull request merged**